Tools/autotest/common.py

```diff
@@ -1467,117 +1467,131 @@ def wait_seconds(self, seconds_to_wait):
+                p = math.degrees(msg.pitch)
+                r = math.degrees(msg.roll)
+                self.progress("Pitch %d Roll %d" % (p, r))
+                return p
+
+        def validator(value2, target2):
+            return math.fabs((value2 - target2 + 180) % 360 - 180) <= accuracy
+
+        self.wait_and_maintain("Pitch", pitch, lambda: get_pitch(timeout), validator=lambda value2, target2: validator(value2, target2), accuracy=accuracy, timeout=timeout)
+
+    def wait_and_maintain(self, value_name, target, target_retreiver, validator=None, accuracy=2.0, maintain_target_time=0, timeout=30, called_function=None):
         tstart = self.get_sim_time()
-        self.progress("Waiting for pitch of %.02f at %s" % (pitch, time.ctime()))
+        duration_start = None
+        mean_value = 0.0
+        last_value = 0.0
+        counter = 0.0
+        self.progress("Waiting for %s %.02f with accuracy %.02f" % (value_name, target, accuracy))
+        last_print_time = 0
         while self.get_sim_time_cached() < tstart + timeout:
-            m = self.mav.recv_match(type='ATTITUDE', blocking=True)
-            p = math.degrees(m.pitch)
-            r = math.degrees(m.roll)
-            self.progress("Pitch %d Roll %d" % (p, r))
-            if math.fabs(p - pitch) <= accuracy:
-                self.progress("Attained pitch %d" % pitch)
-                return True
-        raise WaitPitchTimeout("Failed to attain pitch %d" % pitch)
+            if called_function is not None:
+                called_function(last_value, target)
+            if target_retreiver is not None:
+                last_value = target_retreiver()
+            if self.get_sim_time_cached() - last_print_time > 1:
+                self.progress("%s=%0.2f (want %f +- %f)" %
+                              (value_name, last_value, target, accuracy))
+                last_print_time = self.get_sim_time_cached()
+            if validator is not None:
+                is_value_valid = validator(last_value, target)
+            else:
+                is_value_valid = math.fabs(last_value - target) <= accuracy
+            if is_value_valid:
+                mean_value += last_value
+                counter += 1.0
+                if duration_start is None:
+                    duration_start = self.get_sim_time_cached()
+                if self.get_sim_time_cached() - duration_start >= maintain_target_time:
+                    self.progress("Attained %s=%f" % (value_name, mean_value / counter))
+                    return True
+            else:
+                duration_start = None
+                mean_value = 0.0
+                counter = 0.0
+        raise AutoTestTimeoutException("Failed to attain %s %f, reach %f" % (value_name, target, (mean_value / counter) if counter != 0.0 else last_value))
 
-    def wait_heading(self, heading, accuracy=5, timeout=30):
+    def wait_heading(self, heading, accuracy=5, maintain_target_time=0, timeout=30, called_function=None):
```

Comment:
Doing the `**args` thing here to pass through keywords we don't recognise will make adding functionality to `wait_and_maintain` easier.  Similarly to other frontends to that function.